\documentclass[11pt, a4paper]{article}
%
\usepackage[a4paper, total={6in, 8.5in}]{geometry}
\usepackage{makeidx}
\usepackage{pslatex}
\usepackage{graphicx}
\usepackage{amssymb}
\usepackage{authblk}
\usepackage{titlesec}
\usepackage[obeyspaces]{url}

\setlength{\fboxsep}{8pt}
\setlength{\fboxrule}{0pt}
\setlength{\parindent}{0pt}
\setlength{\parskip}{8pt}

\titlespacing*\section{0pt}{12pt plus 4pt minus 2pt}{4pt plus 2pt minus 2pt}

\begin{document}

\title{GREDU User Reference Manual}

\author{Mingfu Shao \thanks{shaomingfu@gmail.com}}

\affil{Laboratory for Computational Biology and Bioinfomratics, \\
	\'Ecole Polytechnique F\'ed\'erale de Lausanne~(EPFL), \\
		Lausanne, Switzerland}

\maketitle            

\section{Introduction}
GREDU~(Genome REarrangements with DUplications) is a software package
implemented several exact algorithms to compute the edit distances under
various evolutionaly models for two genomes with duplicate genes.
Specifically, GREDU contains the following five programs:
\begin{itemize}
\item[1.] {\bf dcj}, implements an exact algorithm to compute the
DCJ~(Double-Cut-and-Join) distance between two genomes with duplicate
genes~\cite{shao14a}.
\item[2.] {\bf segdcj}, implements an exact algorithm to compute the edit
distance under segmental duplications and DCJ operations for two genomes with
duplicate genes~\cite{shao15c}.
\item[3.] {\bf exemplar}, implements an exact algorithm to compute the exemplar
breakpoint distance for two genomes with duplicate genes~\cite{shao15a}.
\item[4.] {\bf intermediate}, implements an exact algorithm to compute the
intermediate breakpoint distance for two genomes with duplicate genes~\cite{shao16}.
\item[5.] {\bf maxmatching}, implements an exact algorithm to compute the
maximum-matching breakpoint distance for two genomes with duplicate genes~\cite{shao16}.
\end{itemize}

\section{Installation}
To install GREDU, you need to download two libraries~(BOOST and GUROBI),
setup the corresponding environmental variables, and then
compile the source code of GREDU.

\subsection{Install BOOST}
Download BOOST from \url{http://www.boost.org}. Uncompress it
somewhere~(compiling and installing are not necessary). Set environment
variable \url{BOOST_HOME} to indicate the directory of BOOST.
For example, for Unix platforms, add the following
statement to the file \url{~/.bash_profile}:

\fbox{ \parbox{0.9\textwidth}{ \setlength\parindent{19pt}
\url{export BOOST_HOME="/directory/to/your/boost/boost_1_60_0"} }}

\subsection{Install GUROBI}
Download GUROBI from \url{http://www.gurobi.com/} and uncompress the
package somewhere~(compiling and installing are not required).
You need to apply an academic license to use
the full features of GUROBI~(Please refer to the GUROBI documentation for more information.)
After that, set two environment
variables, \url{GUROBI_HOME} and \url{GRB_LICENSE_FILE}, which indicates the directory of GUROBI, and
the location of your license file, respectively.
For example, for Unix platforms, add the following
two statements to the file \url{~/.bash_profile}:

\fbox{ \parbox{0.9\textwidth}{ \setlength\parindent{19pt}
\url{export GUROBI_HOME="/directory/to/your/gurobi/linux64"}

\url{export GRB_LICENSE_FILE="/your/license/gurobi.lic"} }}

\subsection{Compile GREDU}
Get the source code of GREDU through \url{git}:

\fbox{ \parbox{0.9\textwidth}{ \setlength\parindent{19pt}
\url{$git clone git@github.com:shaomingfu/scallop.git .} }}

Compile the libraries, main source code, and tools through:

\fbox{ \parbox{0.9\textwidth}{ \setlength\parindent{19pt}
\url{$lib/build.sh} 

\url{$src/build.sh} 

\url{$tools/build.sh} }}

After that all executable files~(\url{dcj}, \url{segdcj}, \url{exemplar},
\url{intermediate} and \url{maxmatching}) will be present at \url{bin}.

\section{Command line}
All five programs use the same parameters. Take \url{exemplar} as an example:

\fbox{ \parbox{0.9\textwidth}{ \setlength\parindent{19pt}
\url{$./exemplar <genome1> <genome2> <ILP-time-limit>} }}

The first two arguments spcifies two files in which the two genomes are
encoded.  The third argument specifies the time limit~(in seconds) for the
GUROBI solver. A set of input-file examples are provided under \url{bin}.
For program \url{dcj}, you can use \url{human.dcj} and \url{mouse.dcj}
as input files. For other four programs, you can use \url{human.all}
and \url{mouse.all} as input files.

\section{Input Format}
The structure of the genome file is as follows.  A genome contains several
linear or circular chromosomes, and each chromosome consists of a sequence of
genes in the order of their location on the chromosome.  Each chromosome
contain several lines, and each line specifies a gene, containing four fields
separated by spaces.
\begin{itemize}
\item[1.] The first field is a \emph{string}, which species the \emph{identifier} of this gene.
The identifier should be unique for each gene. 
\item[2.] The second field is an \emph{signed integer}, which species the \emph{family} of this gene.
Genes in the same gene family should have the same absolute value.
The orientation of this gene is specified by the sign of this integer.
\item[3.] The third field is a \emph{string}, which species the \emph{chromosome name} of this gene.
\item[4.] The fourth field is a \emph{integer} choosing from $\{1,2\}$, where 1 means this chromosome is linear
and 2 means this chromosome is circular.
\end{itemize}

{\bf NOTE:} Make sure that for program \url{dcj}, for each gene family, the number of genes
in each genome in this gene family are equal~(there is no such requirement for the other
four programs).

\section{Output Format}
A file with name \url{mapping} will be generated in the current
directionary specifying the optimal one-to-one correspondence of the genes in the two genomes,
in which the identifiers of the genes are used. For program \url{segdcj},
an additional file \url{segments} will be generated illstruting the optimal
segmental duplications for each genome.  The optimal edit distance between
the two given genomes will be displayed at the bottom line of the standard
output.

\section{Tools}
If you use the data from Ensembl~\texttt{http://www.ensembl.org}, we have provided
some perl scripts at \texttt{tools/ensembltool} to generate the input files for
these programs from raw downloaded data.

To download gene order data from Ensembl, use the \texttt{customise your download} page.
Take human genome as an example.  First, choose database~(\texttt{Ensembl Genes 74}).  
Then choose dataset~(\texttt{Homo sapiens genes}).  
Second, in the Filters, choose proper chromosomes in the \texttt{REGION} field~(1-22, X and Y),
and choose \texttt{protein coding genes} in the \texttt{GENE} field if necessary.  
Third, in the Attributes, choose \texttt{Ensembl Gene ID}, \texttt{Ensembl Transcript ID},
\texttt{Chromosome Name}, \texttt{Strand}, \texttt{Transcript Start},
\texttt{Transcript End} in the \texttt{GENE} field, and \texttt{Ensembl Protein Family ID(s)}
in the \texttt{PROTEIN DOMAINS} field.  Make sure that these attributes are selected in the same order described above.
Last, come to the \texttt{Results} tab and save them to a CSV file.

After downloading the data, for example, human genome and mouse genome, we can
use the perl script \texttt{tools/ensembltool/build.pl} to transform to the required format:

\fbox{ \parbox{0.9\textwidth}{ \setlength\parindent{19pt}
\url{$./build.pl <human.list> <mouse.list> <human.input> <moust.input>} }}

The first two parameters are the names of the two raw data files, and the last two
parameters are the names of the input files of GREDU with the required
format.  \texttt{build.pl} calls the other three scripts in the
same directory, where \texttt{longest.pl} is to select the
longest transcript for each gene, \texttt{family.pl} is to select
those gene families with the same number of genes in each genome, and
\texttt{join.pl} is to transform those genes in the selected families
to the required format.

{\bf NOTE:} for program~\texttt{dcj}, you have to uncomment a few lines of \texttt{family.pl}
to generate the correct input files~(please follow the instruction on line 47).
This is because \texttt{dcj} requires that for each gene family, exactly the
same number of genes should be given. The modified \texttt{family.pl} will
only keep those gene families with the same gene copy numbers and remove others.

\bibliographystyle{unsrt}
\bibliography{gredu}

\end{document}
